sessiontxn/staleread/provider.go

```diff
@@ -86,10 +89,37 @@ func (p *StalenessTxnContextProvider) OnInitialize(ctx context.Context, tp sessi
 }
 
 // OnStmtStart is the hook that should be called when a new statement started
-func (p *StalenessTxnContextProvider) OnStmtStart(_ context.Context) error {
+func (p *StalenessTxnContextProvider) OnStmtStart(ctx context.Context) error {
+	p.ctx = ctx
 	return nil
 }
 
+// ActivateTxn activates the transaction.
+func (p *StalenessTxnContextProvider) ActivateTxn(_ *sessiontxn.EnterNewTxnType) (kv.Transaction, error) {
+	if p.txn != nil {
+		return p.txn, nil
+	}
+
+	err := p.OnInitialize(p.ctx, sessiontxn.EnterNewTxnDefault)
+	if err != nil {
+		return nil, err
+	}
+
+	txnFuture := p.sctx.GetPreparedTxnFuture()
```

Comment:
You should call `NewStaleTxnWithStartTS` because stale read has some it's own states to set.